### PR TITLE
fix: Rework `useAppBootsrap` and fix deeplinks implementations

### PR DESCRIPTION
**Pull request by Ldoppea**
With previous implementation we had to use `setInitialRoute()` and `setInitialScreen()` to define the initial app state

This was problematic as it was difficult to understand which route should be defined as `initial route` and which one should be defined as `initial screen`

Also the usage of `initialRoute.stack = <some_route>` and `initialScreen.stack = <some_route>` with addition to the presence of `intialRoute.route = routes.stack` made it even more confusing about what is a `root` and what is a `stack`

This complexity was due to the fact that we split the routes declarations between `RootNavigator` and `StackNavigator`. Here again the rule defining where to put each route was unclear

To simplify this we chose to refactor the code to use one single `Navigator` for every routes

This simplifies the API as we only need to use `setInitialRoute()` now, and its parameters only consist of a `route` and a set of `params`

The only exception is the nested `MainAppNavigator` that has been introduced in the `fix: Handle deeplinks on cold start` commit. See the commit description to understand why it is needed
___

TODO:

- [x] Tests on iOS
- [x] Tests on Android
- [x] Document test scenario in a Paper document

### CODE REVIEW

File: src/hooks/useAppBootstrap.functions.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import { getErrorMessage } from '/libs/functions/getErrorMessage'
+
+const log = Minilog('useAppBootstrap.functions')
+
+interface OnboardingParams {
+  onboardUrl: string | null
+  fqdn: string | null
+}
+
+export const parseOnboardingURL = (
```
Comment (acezard) [inline src/hooks/useAppBootstrap.functions.ts:12]: I think as a pure function it should be unit tested

File: src/hooks/useAppBootstrap.functions.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    return undefined
+  }
+}
+
+interface FallbackUrl {
+  mainAppFallbackURL: string | undefined
+  cozyAppFallbackURL: string | undefined
+}
+
+export const parseFallbackURL = (url: string | null): FallbackUrl => {
```
Comment (acezard) [inline src/hooks/useAppBootstrap.functions.ts:46]: I think as a pure function it should be unit tested (2)

File: src/hooks/useAppBootstrap.js [hunk 3]
```diff
@@ -1,72 +1,24 @@
...
-  }
-
-  try {
-    const makeURL = new URL(url)
-    const fallback = makeURL.searchParams.get('fallback')
-    const isHome = makeURL.pathname.split('/')[1] === 'home'
+import {
+  parseFallbackURL,
+  parseOnboardingURL
+} from './useAppBootstrap.functions'
```
Comment (acezard) [inline src/hooks/useAppBootstrap.js:16]: think we should still use aliased import

File: src/hooks/useAppBootstrap.js [hunk 4]
```diff
@@ -1,72 +1,24 @@
...
-  } catch (error) {
-    log.error(
-      `Something went wrong while trying to parse fallback URL data: ${error.message}`
-    )
-    return defaultParse
-  }
-}
+if (devConfig.ignoreLogBox) LogBox.ignoreAllLogs()
 
 export const useAppBootstrap = client => {
```
Comment (acezard) [inline src/hooks/useAppBootstrap.js:20]: just thinking: at some point we should get rid of `useAppBootstrap` and separate the logic like you did in this PR into multiple modules that we can compose in a lifecycle app hook or even during runtime after initial boot if needed

File: src/libs/intents/localMethods.ts [hunk 5]
```diff
@@ -30,7 +30,7 @@ export const asyncLogout = async (client?: CozyClient): Promise<null> => {
   await deleteKeychain()
   await clearCookies()
   await clearData()
-  RootNavigation.reset(routes.stack, { screen: 'welcome' })
+  RootNavigation.reset(routes.welcome, { screen: 'welcome' })
```
Comment (acezard) [inline src/libs/intents/localMethods.ts:33]: note for later: `reset('foo', { screen : 'foo' })` looks weird, we should rewrite reset so the reset route is also automatically the first indexed route. Actually I don't know why it would be otherwise

File: src/screens/home/components/HomeView.js [hunk 6]
```diff
@@ -114,6 +125,35 @@ const HomeView = ({ route, navigation, setLauncherContext, setBarStyle }) => {
     }
   }, [uri, client, route, navigation, session])
 
+  useEffect(
```
Comment (acezard) [inline src/screens/home/components/HomeView.js:128]: I think we should strive to not add any more logic into view components at this point.
Using the pure functions (when possible) with hook-as-a-bridge seem like a good replacement.
Also anyway the logic that I see here should not be triggered by a View in any case (seems higher level in terms of abstraction), but I might have missed something

File: src/App.js [hunk 7]
```diff
@@ -50,10 +50,7 @@ const App = ({ setClient }) => {
 
   useNetService(client)
 
-  const { initialScreen, initialRoute, isLoading } = useAppBootstrap(
-    client,
-    setClient
-  )
+  const { initialRoute, isLoading } = useAppBootstrap(client)
```
Comment (acezard) [inline src/App.js:53]: maybe we could use `initialParams` instead of `initialRoute.params`? I think that would go well with the spirit of this PR

File: src/hooks/useAppBootstrap.functions.ts [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+  if (url === null) {
+    return defaultParse
+  }
+
+  try {
+    const makeURL = new URL(url)
+    const fallback = makeURL.searchParams.get('fallback') ?? undefined
+    const isHome =
+      makeURL.pathname.startsWith('/home') ||
```
Comment (acezard) [inline src/hooks/useAppBootstrap.functions.ts:60]: there are a number of "magic strings" here and I'm not sure I'm too comfortable with that. For instance `"home"` already exists in `routes.js`. But what if tomorrow some other route name replaces `home`? Maybe we should use something like `mainPage` or `defaultPage`, just suggestions, trying to pass an idea

File: src/App.js (hunk 7 above)
Comment (Ldoppea) [reply src/App.js:58]: `initialRoute` and `initialParams` are part of the same concept (i.e. deep links) so I would prefer to have them "linked" into the same `useState` (even if today react is able to handle multiple setState in a single re-render)

However I can split them in the hook's return. Something like:
```js
  return {
    client,
    initialRoute: initialRoute.route,
    initialParams: initialRoute.params
    isLoading
  }
```
But I'm not sure this will improve clarity that much

File: src/hooks/useAppBootstrap.functions.ts (hunk 1 above)
Comment (Ldoppea) [reply src/hooks/useAppBootstrap.functions.ts:12]: I didn't add tests because those are already covered by `useAppBootstrap.spec.js`. I see this extraction more like an implementation detail. But this is pretty subjective. If I add unit tests, those will be kind of duplicates. 
What do you think about this?

File: src/hooks/useAppBootstrap.functions.ts (hunk 8 above)
Comment (Ldoppea) [reply src/hooks/useAppBootstrap.functions.ts]: What about this? https://github.com/cozy/cozy-flagship-app/compare/d65f4daf44c04c41d3ab436de7f31e9def13486c..c15517c19bb895b6b1682f74e176cd27f4d98b8b

I'm not sure of how this can change in the future.

For me the `/home` path in UniversalLinks is about some external app that requests to display `cozy-home`. It shouldn't have to worry about our MainApp concept. But maybe it should?
@Crash-- do you have any thoughts on this?

File: src/hooks/useAppBootstrap.js (hunk 3 above)
Comment (Ldoppea) [reply src/hooks/useAppBootstrap.js]: Fixed in https://github.com/cozy/cozy-flagship-app/pull/550/commits/cd72b2044f0c573367989e1bbbd0dd2f1cad6805

File: src/hooks/useAppBootstrap.js (hunk 4 above)
Comment (Ldoppea) [reply src/hooks/useAppBootstrap.js:19]: Yes, now that I extracted the URL parse logic it would be easy to split this into 2 hooks, one for the bootstrap on cold start, one for the URL interception on runtime.

However I'm not sure if the hideSplashScreen part can be easily removed from the bootstrap hook.

Comment (Ldoppea) [reply src/hooks/useAppBootstrap.js:19]: I propose to do this in another PR if you are ok

File: src/hooks/useAppBootstrap.functions.ts (hunk 8 above)
Comment (Crash--) [reply src/hooks/useAppBootstrap.functions.ts]: I'm not sure to really understand the discussion here, but here is what I can say: 
- We'll be able to specify a new defaultApp (maybe what you call mainapp?) for a cozy. For now, this defaultApp is the Cozy-Home, it'll change in the future. 
- If you have universal link for `links.mycozy.cloud/flagship/toto`, then we should open the `toto` app. If we have link to `/flagship/home`, then we have to open the `home` (this is the slug so cozy-home). If we only have `/flagship`, then we should open the defaultApp.

File: src/libs/intents/localMethods.ts (hunk 5 above)
Comment (Ldoppea) [reply src/libs/intents/localMethods.ts:36]: > we should rewrite reset so the reset route is also automatically the first indexed route

I'm not sure about this. In the NetService you use reset to go back to the callbackRoute that is not necessary the first indexed route

Also you can open the app with the first route being `home` but on logout you want to reset to the `welcome` route.
However i'm not sure why we need the `{ screen: ... }` param here.

File: src/screens/home/components/HomeView.js (hunk 6 above)
Comment (Ldoppea) [reply src/screens/home/components/HomeView.js:128]: > the logic that I see here should not be triggered by a View in any case (seems higher level in terms of abstraction), but I might have missed something

Here I got the same problem you had in the past for the lockscreen.
If the initial route is `drive` we want to display the CozyAppScreen, but we still want to load the HomeScreen in background. We also want the back button to correctly go back to the `home` route.

The solution I found was to navigate to `home` first then the home detects it should open `drive`.

However I agree if you want to split this component into a view/hook organisation. I propose to do it in another PR

File: src/hooks/useAppBootstrap.functions.ts (hunk 8 above)
Comment (Ldoppea) [reply src/hooks/useAppBootstrap.functions.ts]: Yes that was the point: mainapp = defaultapp

My understanding is that today `flagship/home` will open the "main screen" (a.k.a CozyHomeScreen as it is named today)
But in the future we will have to check if `flagship/home` is the default app or if we should open it as a cozy-app. But maybe /home just won't exist on cozys with a different defaultapp?

Comment (Crash--) [reply src/hooks/useAppBootstrap.functions.ts]: > But in the future we will have to check if flagship/home is the default app or if we should open it as a cozy-app (using CozyAppScreen). 

Oh, this is a nice remark, because, yes, /home will still be there even with a different defaultapp. Since this is the slug of an app.

File: src/hooks/useAppBootstrap.functions.ts (hunk 1 above)
Comment (acezard) [reply src/hooks/useAppBootstrap.functions.ts:16]: I agree with you, it can be considered an implementation detail and thus shouldn't be tested.

File: src/hooks/useAppBootstrap.functions.ts (hunk 2 above)
Comment (acezard) [reply src/hooks/useAppBootstrap.functions.ts:50]: Edit: implem detail so no need

File: src/hooks/useAppBootstrap.js (hunk 4 above)
Comment (acezard) [reply src/hooks/useAppBootstrap.js:19]: Yes it was definitely for another PR

File: src/libs/intents/localMethods.ts (hunk 5 above)
Comment (acezard) [reply src/libs/intents/localMethods.ts:36]: Well that's the point, with the new routing, reset() doesn't need two params, but just one, since we don't have two navigators (well we still do, but one has only one route so we can consider it as a route).

But anyway this is to be considered, it's not directly linked to your PR but the question is raised from the new routing

File: src/screens/home/components/HomeView.js (hunk 6 above)
Comment (acezard) [reply src/screens/home/components/HomeView.js:128]: Yeah agree for the render problem.

Ok for refactoring later

File: src/App.js (hunk 7 above)
Comment (acezard) [reply src/App.js:58]: alright agree with you

File: src/hooks/useAppBootstrap.functions.ts (hunk 8 above)
Comment (acezard) [reply src/hooks/useAppBootstrap.functions.ts]: yeah that's why I think we should use semantic variables right now as it's very easily done, just to keep in mind that nothing is hardcoded nor it should be

**Review by acezard — APPROVED** [review]
